File: src/main/java/kr/co/lunatalk/domain/order/domain/Order.java
package kr.co.lunatalk.domain.order.domain;

import jakarta.persistence.*;
import kr.co.lunatalk.domain.common.domain.BaseTimeEntity;
import kr.co.lunatalk.domain.delivery.domain.Delivery;
import kr.co.lunatalk.domain.member.domain.Member;
import kr.co.lunatalk.domain.payment.domain.Payment;
import lombok.AccessLevel;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.util.ArrayList;
import java.util.List;

@Entity
@Table(name = "orders")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class Order extends BaseTimeEntity {

	@Id @GeneratedValue(strategy = GenerationType.IDENTITY)
	private Long id;

	@Column(nullable = false, unique = true)
	private String orderNumber;

	@ManyToOne(fetch = FetchType.LAZY)
	@JoinColumn(name = "member_id")
	private Member member;

	private Long totalPrice;

	@Enumerated(EnumType.STRING)
	private OrderStatus status;

	@OneToOne(mappedBy = "order", cascade = CascadeType.ALL)
	private Payment payment;

	@OneToMany(mappedBy = "order", cascade = CascadeType.ALL, orphanRemoval = true)
	private List<OrderItem> orderItems = new ArrayList<>();

	@OneToMany(mappedBy = "order", cascade = CascadeType.ALL)
	private List<Delivery> deliveries = new ArrayList<>();

	@Builder(access = AccessLevel.PRIVATE)
	public Order(String orderNumber, Member member, Long totalPrice, OrderStatus status) {
		this.orderNumber = orderNumber;
		this.member = member;
		this.totalPrice = totalPrice;
		this.status = status;
	}

	public static Order createOrder(String orderNumber, Member member, Long totalPrice) {
		return Order.builder()
			.orderNumber(orderNumber)
			.member(member)
			.totalPrice(totalPrice)
			.status(OrderStatus.CREATED)
			.build();
	}

	public void addOrderItem(OrderItem orderItem) {
		orderItems.add(orderItem);
	}

	public void updateTotalPrice(long totalPrice) {
		this.totalPrice = totalPrice;
	}

	public void updateStatus(OrderStatus status) {
		this.status = status;
	}
}

File: src/main/kotlin/kr/co/lunatalk/domain/order/domain/Order.kt
package kr.co.lunatalk.domain.order.domain

import jakarta.persistence.*
import kr.co.lunatalk.domain.common.domain.BaseTimeEntity
import kr.co.lunatalk.domain.delivery.domain.Delivery
import kr.co.lunatalk.domain.member.domain.Member

@Entity
@Table(name = "orders")
open class Order protected constructor() : BaseTimeEntity() {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    open var id: Long? = null
        protected set

    @Column(nullable = false, unique = true)
    open var orderNumber: String? = null
        protected set

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "member_id")
    open var member: Member? = null
        protected set

    open var totalPrice: Long? = null
        protected set

    @Column(length = 30)
    @Enumerated(EnumType.STRING)
    open var status: OrderStatus? = null
        protected set

    @OneToMany(mappedBy = "order", cascade = [CascadeType.ALL], orphanRemoval = true)
    open var orderItems: MutableList<OrderItem> = mutableListOf()
        protected set

    @OneToMany(mappedBy = "order", cascade = [CascadeType.ALL])
    open var deliverys: MutableList<Delivery> = mutableListOf()
        protected set

    private constructor(
        orderNumber: String?,
        member: Member?,
        totalPrice: Long?,
        status: OrderStatus?
    ) : this() {
        this.orderNumber = orderNumber
        this.member = member
        this.totalPrice = totalPrice
        this.status = status
    }

    fun addOrderItem(orderItem: OrderItem) {
        orderItems.add(orderItem)
    }

    fun updateTotalPrice(totalPrice: Long) {
        this.totalPrice = totalPrice
    }

    fun updateStatus(status: OrderStatus) {
        this.status = status
    }

    companion object {
        fun createOrder(orderNumber: String, member: Member, totalPrice: Long): Order {
            return Order(
                orderNumber = orderNumber,
                member = member,
                totalPrice = totalPrice,
                status = OrderStatus.ORDERED
            )
        }
    }
}
